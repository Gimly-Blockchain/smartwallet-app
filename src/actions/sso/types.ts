import { PublicProfileClaimMetadata } from 'cred-types-jolocom-core/types'

/**
 * @dev Simply using all claims required by the public profile
 */

type IssuerPublicProfileSummary = PublicProfileClaimMetadata['claimInterface']

/**
 * @dev An identity summary is composed of a DID + all public info (currently public profile)
 */
export interface IdentitySummary {
  did: string
  publicProfile?: IssuerPublicProfileSummary
}
