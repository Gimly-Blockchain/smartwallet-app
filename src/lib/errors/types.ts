import { routeList } from 'src/routeList'
import { ErrorCode } from './codes'

export { ErrorCode }

export interface IAppError {
  origError: any
  navigateTo: routeList
}

export interface UserReport {
  userError: string | undefined
  userDescription: string
  userContact: string
  sendPrivateData: boolean
}

export interface ErrorReport extends UserReport {
  error: IAppError | Error | undefined
}
