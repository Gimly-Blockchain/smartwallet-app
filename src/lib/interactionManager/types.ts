
import { JSONWebToken, JWTEncodable } from 'jolocom-lib/js/interactionTokens/JSONWebToken'


/* Responsabilities:
 * 1. continue an interaction
 */
export interface IInteractionManager {
    tokenRecieved: (token: JSONWebToken<JWTEncodable>) => void

}
