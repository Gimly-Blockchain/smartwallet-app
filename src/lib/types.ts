import { JSONWebToken, JWTEncodable } from 'jolocom-lib/js/interactionTokens/JSONWebToken'

export type SendFn = (token: JSONWebToken<JWTEncodable>) => Promise<any>
