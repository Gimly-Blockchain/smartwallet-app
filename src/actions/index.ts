import * as accountActions from 'src/actions/account/'
import * as registrationActions from 'src/actions/registration/'
import * as navigationActions from 'src/actions/navigation/'
import * as genericActions from 'src/actions/generic/'
import * as ssoActions from 'src/actions/sso/'
import * as interactionHandlerActions from 'src/actions/interactionHandler'
import * as paymentActions from 'src/actions/payment/'

export {
  accountActions,
  registrationActions,
  navigationActions,
  genericActions,
  ssoActions,
  interactionHandlerActions,
  paymentActions
}
