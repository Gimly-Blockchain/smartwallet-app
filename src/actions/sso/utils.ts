import { IssuerPublicProfileSummary } from './types'
import { Identity } from 'jolocom-lib/js/identity/identity'

export const parsePublicProfile = (identity: Identity) => {
  const { publicProfile, did } = identity
  if (!publicProfile) {
    return {
      did,
    }
  }
  const { id, ...parsedProfile } = publicProfile.claim
  return {
    did,
    publicProfile: parsedProfile as IssuerPublicProfileSummary,
  }
}
