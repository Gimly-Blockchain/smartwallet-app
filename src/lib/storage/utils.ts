// TODO rework
// Multiple credentials of the same type, i.e. ['Credential', 'ProofOfNameCredential']
import {
  CredentialEntity,
  VerifiableCredentialEntity,
} from './entities'

interface ModifiedCredentialEntity {
  propertyName: string
  propertyValue: string[]
  verifiableCredential: VerifiableCredentialEntity
}

export const groupAttributesByCredentialId = (
  credentials: CredentialEntity[],
): ModifiedCredentialEntity[] => {
  // Convert values to arrays for easier concatination later
  const modifiedAttributes = credentials.map(credential => ({
    ...credential,
    propertyValue: [credential.propertyValue], // "eugeniu@jolocom.com" -> ["eugeniu@jolocom.com"]
  }))

  // Helper function
  const findByCredId = (
    arrToSearch: ModifiedCredentialEntity[],
    value: ModifiedCredentialEntity,
  ) =>
    arrToSearch.findIndex(
      entry => entry.verifiableCredential.id === value.verifiableCredential.id,
    )

  debugger
  return modifiedAttributes.reduce(
    (acc: ModifiedCredentialEntity[], curr: ModifiedCredentialEntity) => {
      const matchingIndex = findByCredId(acc, curr)

      // This seems to be a hack, uniting individual claims from the credential back together
      if (matchingIndex >= 0) {
        acc[matchingIndex].propertyValue = [
          ...acc[matchingIndex].propertyValue,
          ...curr.propertyValue,
        ]
        return acc
      } else {
        return [...acc, curr]
      }
    },
    [],
  )
}
