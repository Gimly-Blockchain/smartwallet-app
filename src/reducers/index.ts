import { combineReducers } from 'redux'
import { settingsReducer, SettingsState } from 'src/reducers/settings/'
import { accountReducer, AccountState } from 'src/reducers/account/'
import {
  registrationReducer,
  RegistrationState,
} from 'src/reducers/registration/'
import { navigationReducer } from 'src/reducers/navigation/'
import { NavigationState } from 'react-navigation'
import { ssoReducer, SsoState } from 'src/reducers/sso/'

export const rootReducer = combineReducers({
  settings: settingsReducer,
  account: accountReducer,
  registration: registrationReducer,
  navigation: navigationReducer,
  sso: ssoReducer,
})

export interface RootState {
  readonly settings: SettingsState
  readonly account: AccountState
  readonly registration: RegistrationState
  readonly navigation: NavigationState
  readonly sso: SsoState
}
