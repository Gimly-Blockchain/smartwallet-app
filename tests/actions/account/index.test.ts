import { accountActions } from 'src/actions/'
import data from '../registration/data/mockRegistrationData'
import { JolocomLib } from 'jolocom-lib'
import { RootState } from 'src/reducers'
import { createMockStore, stub } from 'tests/utils'
import { DidDocument } from 'jolocom-lib/js/identity/didDocument/didDocument'
import { IContractsAdapter, IContractsGateway } from 'jolocom-lib/js/contracts/types';

describe('Account action creators', () => {
  const initialState: Partial<RootState> = {
    registration: {
      loading: {
        loadingMsg: '',
        isRegistering: false,
      },
    },
    account: {
      loading: false,
      claims: {
        selected: {
          credentialType: 'Email',
          claimData: {
            email: 'test@test.com',
          },
          id: '',
          issuer: {
            did: 'did:jolo:test',
          },
          subject: 'did:jolo:test',
        },
        pendingExternal: {
          offer: [],
          offeror: {
            did: '',
          },
        },
        decoratedCredentials: {},
      },
      did: {
        did: '',
      },
    },
  }

  const mockIdentityWallet = {
    identity: { did: 'did:jolo:first', didDocument: {} },
    didDocument: {},
  }

  const mockMiddleware = {
    registry: {
      contractsAdapter: stub<IContractsAdapter>(),
      contractsGateway: stub<IContractsGateway>(),
      authenticate: jest.fn().mockResolvedValue(mockIdentityWallet),
    },
    storageLib: {
      get: {
        persona: jest.fn().mockResolvedValue({
          did: 'did:jolo:first',
        }),
        encryptedSeed: jest.fn().mockResolvedValue('mockencryptedvalue'),
        didDoc: jest.fn().mockResolvedValue(undefined),
      },
      store: {
        didDoc: jest.fn().mockResolvedValue(undefined),
      },
    },
    keyChainLib: {
      getPassword: jest.fn().mockResolvedValue('secret'),
    },
    encryptionLib: {
      decryptWithPass: jest.fn().mockReturnValue('a'.repeat(64)),
    },
    identityWallet: mockIdentityWallet,
  }

  const mockStore = createMockStore(initialState, mockMiddleware)

  beforeEach(mockStore.reset)

  it('Should correctly handle existing user identity if not cached', async () => {
    await mockStore.dispatch(accountActions.checkIdentityExists)
    expect(mockMiddleware.registry.authenticate).toHaveBeenCalledTimes(1)
    expect(mockMiddleware.storageLib.store.didDoc).toHaveBeenCalledWith(
      mockMiddleware.identityWallet.identity.didDocument,
    )

    expect(mockStore.getActions()).toMatchSnapshot()

    // TODO(@mnzaki)
    // mockStore.reset() should handle this, probably by storing references to
    // all jest.fn that are part of mockMiddleware, on stub creation
    mockMiddleware.registry.authenticate.mockReset()
  })

  it('Should correctly handle existing user identity if cached', async () => {
    const mockDidDoc = DidDocument.fromJSON({
      id: 'did:jolo:first',
      publicKey: [
        //@ts-ignore
        {
          publicKeyHex: '0xabc',
          id: 'did:jolo:first#123',
        },
      ],
    })

    mockStore.backendMiddleware.storageLib.get.didDoc = jest
      .fn()
      .mockResolvedValue(mockDidDoc)

    await mockStore.dispatch(accountActions.checkIdentityExists)
    expect(mockMiddleware.registry.authenticate).not.toHaveBeenCalled()
    expect(mockMiddleware.identityWallet.identity.didDocument).toStrictEqual(
      mockDidDoc,
    )
    expect(mockStore.getActions()).toMatchSnapshot()
  })

  it('Should correctly handle more existing user identities', async () => {
    mockMiddleware.storageLib.get.persona.mockResolvedValueOnce([
      { did: 'did:jolo:first' },
    ])
    await mockStore.dispatch(accountActions.checkIdentityExists)
    expect(mockStore.getActions()).toMatchSnapshot()
  })

  it('Should correctly handle an empty encrypted seed table', async () => {
    mockMiddleware.storageLib.get.persona.mockResolvedValueOnce([])
    mockMiddleware.storageLib.get.encryptedSeed.mockReturnValueOnce(null)
    await mockStore.dispatch(accountActions.checkIdentityExists)
    expect(mockStore.getActions()).toMatchSnapshot()
  })

  it('Should correctly retrieve claims from device storage db on setClaimForDid', async () => {
    const { identityWallet, testSignedCredentialDefault } = data

    const backendMiddleware = {
      storageLib: {
        get: {
          verifiableCredential: jest
            .fn()
            .mockResolvedValue([
              JolocomLib.parse.signedCredential(testSignedCredentialDefault),
            ]),
          credentialMetadata: jest.fn().mockResolvedValue({}),
          publicProfile: jest.fn().mockResolvedValue({}),
        },
      },
      identityWallet,
    }

    const altMockStore = createMockStore(initialState, backendMiddleware)

    await altMockStore.dispatch(accountActions.setClaimsForDid)
    expect(altMockStore.getActions()).toMatchSnapshot()
  })

  it('should correctly save claim', async () => {
    const { identityWallet } = data

    const backendMiddleware = {
      keyChainLib: {
        getPassword: jest.fn().mockResolvedValue('sekrit'),
      },
      storageLib: {
        store: {
          verifiableCredential: jest.fn().mockResolvedValue([]),
        },
        get: {
          verifiableCredential: jest.fn().mockResolvedValue([]),
          publicProfile: jest.fn().mockReturnValue({}),
        },
      },
      identityWallet,
    }

    const altMockStore = createMockStore(initialState, backendMiddleware)

    await altMockStore.dispatch(accountActions.saveClaim)
    expect(altMockStore.getActions()).toMatchSnapshot()
  })
})
