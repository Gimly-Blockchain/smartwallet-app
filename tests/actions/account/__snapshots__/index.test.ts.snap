// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account action creators Should correctly handle an empty encrypted seed table 1`] = `
Array [
  Object {
    "actions": Array [
      Object {
        "routeName": "Landing",
        "type": "Navigation/NAVIGATE",
      },
    ],
    "index": 0,
    "key": undefined,
    "type": "Navigation/RESET",
  },
]
`;

exports[`Account action creators Should correctly handle more existing user identites 1`] = `
Array [
  Object {
    "type": "DID_SET",
    "value": "did:jolo:first",
  },
  Object {
    "actions": Array [
      Object {
        "routeName": "Home",
        "type": "Navigation/NAVIGATE",
      },
    ],
    "index": 0,
    "key": undefined,
    "type": "Navigation/RESET",
  },
]
`;

exports[`Account action creators Should correctly handle one existing user identity 1`] = `
Array [
  Object {
    "type": "DID_SET",
    "value": "did:jolo:mock",
  },
  Object {
    "actions": Array [
      Object {
        "routeName": "Home",
        "type": "Navigation/NAVIGATE",
      },
    ],
    "index": 0,
    "key": undefined,
    "type": "Navigation/RESET",
  },
]
`;

exports[`Account action creators Should correctly retrieve claims from device storage db on setClaimForDid 1`] = `
Array [
  Object {
    "claims": Object {
      "Contact": Array [
        Object {
          "claimData": Object {
            "email": "test@jolocom.com",
          },
          "credentialType": "Email",
          "expires": 2019-10-15T19:59:38.406Z,
          "id": "claimId:61adc6e7b1448",
          "issuer": Object {
            "did": "did:jolo:b310d293aeac8a5ca680232b96901fe85988fde2860a1a5db69b49762923cc88",
          },
          "subject": "did:jolo:b310d293aeac8a5ca680232b96901fe85988fde2860a1a5db69b49762923cc88",
        },
      ],
    },
    "type": "SET_CLAIMS_FOR_DID",
  },
]
`;

exports[`Account action creators should correctly save claim 1`] = `
Array [
  Object {
    "claims": Object {},
    "type": "SET_CLAIMS_FOR_DID",
  },
  Object {
    "actions": Array [
      Object {
        "routeName": "Home",
        "type": "Navigation/NAVIGATE",
      },
    ],
    "index": 0,
    "key": undefined,
    "type": "Navigation/RESET",
  },
]
`;
